### PR TITLE
U/ejeffrey/late unflatten

### PR DESCRIPTION
**Pull request by ejeffrey**
This patch extends the use of the FlatData construct and allows for server settings to decline to have its arguments unflattened.

The idea here is that servers that just accept a bunch of data and pass it on to other servers can avoid the unflattening/flattening step altogether.

Currently this is only implemented for requests.  Before merging I want to do the same for responses.  Just marking this so that @maffoo can look at it.

### CODE REVIEW

**Comment by ejeffrey** [comment]
OK, response data is now able to be received as raw flatdata.  The way you do this is to use "unflatten=False" as a keyword argument either to the setting wrapper `__call__` or to the `packet.send()` method.

File: labrad/backend.py [hunk 1]
```diff
@@ -390,6 +390,7 @@ def handleResponse(self, _source, _context, request, records):
         future = self.requests[n]
         del self.requests[n]
         self.pool.add(n)
+        records = [ (ID, flat_data.unflatten()) for ID, flat_data in flat_records ]
```
Comment (maffoo) [inline labrad/backend.py]: pep8: [no whitespace inside brackets](https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements)

File: labrad/decorators.py [hunk 2]
```diff
@@ -21,202 +21,140 @@
...
 def _product(lists):
     """Return the cartesian product of a list of lists."""
     if not len(lists): return [[]]
     return [[h] + t for h in lists[0] for t in _product(lists[1:])]
 
-class Setting(object):
-    def __init__(self, func):
-        self.func = func
+def setting(lr_ID, lr_name=None, returns=[], unflatten=True, **params):
+    def decorator(func):
```
Comment (maffoo) [inline labrad/decorators.py]: This should use functools.wraps to preserve the docstring etc. on the server method:
```python
@functools.wraps(func)
def decorator(func):
    ...
```

File: labrad/decorators.py [hunk 3]
```diff
@@ -21,202 +21,140 @@
...
-    def getRegistrationInfo(self):
-        return (long(self.ID), self.name, self.description,
-                self.accepts, self.returns, self.notes)
+class Setting(object):
+    """
+    Represent a single setting within a Server object.
 
-    def handleRequest(self, c, data):
-        return self.func(c, data)
+    args:
```
Comment (maffoo) [inline labrad/decorators.py]: Please follow the formatting conventions in the [google python style guide](https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments) and include types:
```python
Args:
    func (callable): Function to implement the setting.
    ...
```

File: labrad/decorators.py [hunk 4]
```diff
@@ -21,202 +21,140 @@
...
-
 def _product(lists):
     """Return the cartesian product of a list of lists."""
     if not len(lists): return [[]]
     return [[h] + t for h in lists[0] for t in _product(lists[1:])]
 
-class Setting(object):
-    def __init__(self, func):
-        self.func = func
+def setting(lr_ID, lr_name=None, returns=[], unflatten=True, **params):
```
Comment (maffoo) [inline labrad/decorators.py]: This is the method people are going to call, rather than the `Setting` class, so please add a doc string, even if just to point to the docstring on the `Setting` class.

File: labrad/decorators.py [hunk 5]
```diff
@@ -21,202 +21,140 @@
...
+    returns:    The possible labrad type strings returned by the setting
 
-    The only required parameter is an integer ID.  You may
-    also provide a name to override the name of the decorated
-    function.  In addition, accepted types for each of the setting
-    parameters may be provided as named parameters with a list of
-    strings of allowed types.
+    **kw:       Additional keyword arguments indicate arguments to the setting.  Each keyword
+                should be associated with one or more labrad type strings indicating data types
+                it accepts.
```
Comment (maffoo) [inline labrad/decorators.py]: "associated with" is a little vague, and it's also not clear that the names of these arguments should match the names of arguments to the server method being decorated. I would suggest rewording to make this clear.

File: labrad/decorators.py [hunk 6]
```diff
@@ -21,202 +21,140 @@
...
+    lr_ID:      The ID number of the setting used in the labrad protocol
+    lr_name:    The setting name.  By default, this is derived from the function name
+    returns:    The possible labrad type strings returned by the setting
 
-    The only required parameter is an integer ID.  You may
-    also provide a name to override the name of the decorated
-    function.  In addition, accepted types for each of the setting
-    parameters may be provided as named parameters with a list of
-    strings of allowed types.
+    **kw:       Additional keyword arguments indicate arguments to the setting.  Each keyword
```
Comment (maffoo) [inline labrad/decorators.py]: On `__init__` this is actually called `**params`, rather than `**kw`.

File: labrad/decorators.py [hunk 7]
```diff
@@ -21,202 +21,140 @@
...
+        #    if all other arguments are optional.
+        # 4) If more than one argument are required, we expect to always receive
+        #    a tuple and unpack it
+        # 5) If only a single argument is allowed, we never unpack tuples
+        # 6) If both =1 and >1 arguments are allowed, it is ambiguous whether to
+        #    unpack tuples, so this case is not allowed:  The first argument
+        #    cannot be a tuple or '?' tag if the second argument is optional.
+        
+        argspec = inspect.getargspec(self.func)
+        args = argspec.args[2:]
```
Comment (maffoo) [inline labrad/decorators.py]: A comment about where the 2 comes from would be helpful (first two args are self and context object).

File: labrad/decorators.py [hunk 8]
```diff
@@ -21,202 +21,140 @@
...
+
+        if inspect.isgeneratorfunction(func):
+            self.func = defer.inlineCallbacks(func)
```
Comment (maffoo) [inline labrad/decorators.py]: I would put this above when you actually assign self.func. Here it's sort of mixed in with the argument validation.

File: labrad/decorators.py [hunk 9]
```diff
@@ -21,202 +21,140 @@
...
+
+        for p in params.keys():
             if p not in args:
-                raise Exception("'%s' is not a valid parameter." % p)
-            # turn single string annotations into lists
-            if isinstance(params[p], str):
+                raise ValueError("Setting parameter {} not accepted by function".format(p))
+            if isinstance(params[p], basestring):
                 params[p] = [params[p]]
+            #params[p] = [T.parseTypeTag(tag) for tag in params[p]]
```
Comment (maffoo) [inline labrad/decorators.py]: remove commented code.

File: labrad/decorators.py [hunk 10]
```diff
@@ -21,202 +21,140 @@
...
+            self.unpack = "always"
+        elif Nparams > 1:
+            self.unpack = "optional"
+        else: # one or fewer arguments
+            self.unpack = "never"
+
+        self.allow_none = Nrequired == 0
+
+        if Nparams:
+            for tag_str in params[args[0]]:
```
Comment (maffoo) [inline labrad/decorators.py]: I've tried to be consistent elsewhere in calling the string form a "tag" and the parsed object a "type". of course, the name "type" is a name of a builtin class, so usually "t" is used for that. It'd be nice to stay consistent with that convention. (One place where I know we're inconsistent is in the FlatData object which has a "tag" field that contains the parsed type object; we might want to fix that, too.)

File: labrad/decorators.py [hunk 11]
```diff
@@ -21,202 +21,140 @@
...
+                tag = T.parseTypeTag(tag_str)
+                if isinstance(tag, T.LRNone) and Nrequired != 1:
+                    raise ValueError("First argument %s cannot accept '' unless other arguments are optional" % (args[0],))
+                if isinstance(tag, (T.LRAny, T.LRCluster)) and self.unpack=="optional":
```
Comment (maffoo) [inline labrad/decorators.py]: nit: spaces around `==`

File: labrad/decorators.py [hunk 12]
```diff
@@ -21,202 +21,140 @@
...
+        else: # one or fewer arguments
+            self.unpack = "never"
+
+        self.allow_none = Nrequired == 0
+
+        if Nparams:
+            for tag_str in params[args[0]]:
+                tag = T.parseTypeTag(tag_str)
+                if isinstance(tag, T.LRNone) and Nrequired != 1:
+                    raise ValueError("First argument %s cannot accept '' unless other arguments are optional" % (args[0],))
```
Comment (maffoo) [inline labrad/decorators.py]: Use `.format` instead of `%`. Also, line is too long

File: labrad/decorators.py [hunk 13]
```diff
@@ -21,202 +21,140 @@
...
+                if isinstance(tag, (T.LRAny, T.LRCluster)) and self.unpack=="optional":
+                    raise ValueError("First argument %s cannot accept type %s because other arguments are optional" % (args[0], tag))
+            for arg in args[1:]:
+                for tag in arg:
```
Comment (maffoo) [inline labrad/decorators.py]:
```python
for tag in params[arg]:
    ...
```

File: labrad/decorators.py [hunk 14]
```diff
@@ -21,202 +21,140 @@
...
+                    if isinstance(tag, T.LRNone):
+                        raise ValueError("Argument %s cannot accept ''" % (arg,))
```
Comment (maffoo) [inline labrad/decorators.py]: use `.format`

File: labrad/decorators.py [hunk 15]
```diff
@@ -21,202 +21,140 @@
...
+
+        # Build list of accepted data types.
+        # This is basically every combination of accepted types for each agrument,
+        # including omitting any number of trailing optional arguments.
+
+        accepted_types = []
+        for ii in range(Nrequired, Nparams+1):
```
Comment (maffoo) [inline labrad/decorators.py]: why `ii` instead of `i`?

File: labrad/decorators.py [hunk 16]
```diff
@@ -21,202 +21,140 @@
...
-                if args[0] not in params:
-                    raise Exception('Must specify types for first argument '
-                                    'when fewer than two args are required.')
-                for s in params[args[0]]:
-                    t = T.parseTypeTag(s)
-                    if isinstance(t, (T.LRAny, T.LRCluster)):
-                        raise Exception('Cannot accept cluster or ? in first '
-                                        'arg when fewer than two args are '
-                                        'required.')
+                accept_tuples = itertools.product(*[params[arg] for arg in args[:ii] ])
```
Comment (maffoo) [inline labrad/decorators.py]: remove extra space before closing bracket

File: labrad/decorators.py [hunk 17]
```diff
@@ -21,202 +21,140 @@
...
+                accepted_types.extend( ''.join(x) for x in accept_tuples)
```
Comment (maffoo) [inline labrad/decorators.py]: remove space after opening paren

File: labrad/decorators.py [hunk 18]
```diff
@@ -21,202 +21,140 @@
...
-
-                @wraps(f)
-                def handleRequest(self, c, data):
-                    if isinstance(data, tuple):
-                        return f(self, c, *data)
-                    elif data is None:
-                        return f(self, c)
-                    else:
-                        return f(self, c, data)
+        if self.unpack == "always" or self.unpack=="optional" and isinstance(data_type, T.LRCluster):
```
Comment (maffoo) [inline labrad/decorators.py]: nit: space around `==`

File: labrad/decorators.py [hunk 19]
```diff
@@ -21,202 +21,140 @@
...
+            if (self.unflatten):
```
Comment (maffoo) [inline labrad/decorators.py]: no parens around if condition

File: labrad/decorators.py [hunk 20]
```diff
@@ -21,202 +21,140 @@
...
-                    if data is None:
-                        return f(self, c)
-                    return f(self, c, data)
-
+        if Nrequired > 1:
+            self.unpack = "always"
+        elif Nparams > 1:
+            self.unpack = "optional"
+        else: # one or fewer arguments
+            self.unpack = "never"
```
Comment (maffoo) [inline labrad/decorators.py]: `unpack` sounds a lot like `unflatten`. Maybe we could call this something like `expand_cluster` or something to indicate that this controls whether an incoming cluster is expanded into multiple function args.

File: labrad/protocol.py [hunk 21]
```diff
@@ -170,7 +171,10 @@ def clearCache(self):
         self._serverCache = {}
         self._settingCache = {}
 
-    def _sendRequestNoLookup(self, target, records, context=(0, 0), timeout=None):
+    def unflattenResponse(self, response_records):
+        return [ (ID, data.unflatten()) for (ID, data) in response_records ]
```
Comment (maffoo) [inline labrad/protocol.py]: no space inside brackets

File: labrad/protocol.py [hunk 22]
```diff
@@ -186,6 +190,8 @@ def _sendRequestNoLookup(self, target, records, context=(0, 0), timeout=None):
                                             errors.RequestTimeoutError())
             d.addBoth(self._cancelTimeout, timeoutCall)
         d.addBoth(self._finishRequest, n)
+        if(unflatten):
```
Comment (maffoo) [inline labrad/protocol.py]: no parens on if condition

File: labrad/protocol.py [hunk 23]
```diff
@@ -270,7 +277,7 @@ def _dispatchMessage(self, source, context, records):
                                          msgCtx, data, listener)
                         else:
                             yield d
-                    except Exception:
+                    except Exception as e:
```
Comment (maffoo) [inline labrad/protocol.py]: is `e` used anywhere?

File: labrad/stream.py [hunk 24]
```diff
@@ -42,10 +42,13 @@ def unflattenRecords(data, endianness='>'):
     s = T.Buffer(data)
     while len(s):
         ID, tag, data = T.unflatten(s, RECORD_TYPE, endianness)
-        rec = ID, T.unflatten(data, tag, endianness)
+        rec = ID, T.FlatData(data, T.parseTypeTag(tag), endianness)
         records.append(rec)
     return records
 
+def unflattenRecord(flatdata, endianness):
```
Comment (maffoo) [inline labrad/stream.py]: Looks like this function is not used. Is it needed?

File: labrad/types.py [hunk 25]
```diff
@@ -295,8 +295,20 @@ def unflatten(s, t, endianness='>'):
     return t.__unflatten__(s, endianness)
 
 
-FlatData = collections.namedtuple('FlatData', ['bytes', 'tag'])
+class FlatData(collections.namedtuple('FlatDataBase', ['bytes', 'tag', 'endianness'])):
+    """FlatData represents unflattened data.
 
+    The elements are
```
Comment (maffoo) [inline labrad/types.py]: Style guide:
```
Attributes:
    bytes (bytes): ...
```

File: labrad/types.py [hunk 26]
```diff
@@ -295,8 +295,20 @@ def unflatten(s, t, endianness='>'):
...
+
+    bytes:      The data as intended to be transmitted over the wire
+    tag:        The LabRAD data type in the form of an LRType object.
```
Comment (maffoo) [inline labrad/types.py]: As I mentioned above, it would be good here to align with the distinction between tag (string) and type (LRType), but not sure the right way to do that, while keeping the `namedtuple` semantics. Maybe keep the parsed type as a `.t` field, and make `tag` be a proprty that returns a string?

File: labrad/types.py [hunk 27]
```diff
@@ -936,6 +953,13 @@ def __unflatten__(self, s, endianness):
         """Unflatten items into a python tuple."""
         return tuple(unflatten(s, t, endianness) for t in self.items)
 
+    def partial_unflatten(self, s, endianness):
+        """Unflatten a cluster into a tuple of FlatData objects"""
+        s = Buffer(s)
+        s2 = Buffer(s)
+        item_sizes = [item.__width__(s, endianness) for item in self.items]
+        return [FlatData(s2.get(item_size), item, endianness) for (item, item_size) in zip(self.items, item_sizes)]
```
Comment (maffoo) [inline labrad/types.py]: line length. Also, docstring says it returns tuple, but actually return list.

**Comment by maffoo** [comment]
This is great. I made a number of small, mostly syntactic comments. Would be nice to fix the tag/type distinction, but could also defer that to later.

File: labrad/decorators.py (hunk 2 above)
Comment (ejeffrey) [inline labrad/decorators.py]: Note that the decorator function doesn't actually become the server method.

File: labrad/decorators.py (hunk 8 above)
Comment (ejeffrey) [inline labrad/decorators.py]: This has to go after we pull out the argspec, since generators don't have an argspec.  The alternative was to call getargspec on 'func' and have that be different from self.func but I think that is to unclear -- someone might see that and change it to self.func for consistency.

File: labrad/decorators.py (hunk 2 above)
Comment (maffoo) [inline labrad/decorators.py]: Right; so many levels of indirection! :)

I guess `Setting` still needs a `__call__` and that's what should wrap `func`.

File: labrad/decorators.py [hunk 28]
```diff
@@ -21,202 +21,167 @@
...
+    lr_ID (int):       The ID number of the setting used in the labrad protocol
+    lr_name (str):     The setting name.  By default, this is derived from the
+                       function name.
+    returns:           The labrad type tag or list of tags the setting returns
+    unflatten (bool):  Request automatic unflattening of incoming data.
+                       (default True)
+    **params:          Additional keyword arguments indicate arguments to the
+                       setting.  Each keyword should be a string matching one
+                       of the formal parameters of func, while the value is
+                       the type tag or list of tags accepted by that parameter.
```
Comment (maffoo) [inline labrad/decorators.py]: Should put the constructor arg docs on the constructor docstring instead of class docstring. Also, nit: args should be indented within the `Args:` section:
```
Args:
    func (callable): Function to implement the setting.
    ...
```

File: labrad/decorators.py [hunk 29]
```diff
@@ -21,202 +21,167 @@
...
-                    elif data is None:
-                        return f(self, c)
-                    else:
-                        return f(self, c, data)
+            return self
+        
+    def handleRequest(self, server, c, flat_data):
+        data_type = flat_data.tag
+
+        if self.expand_cluster == "always" or self.expand_cluster == "optional" and isinstance(data_type, T.LRCluster):
```
Comment (maffoo) [inline labrad/decorators.py]: line too long.

File: labrad/types.py [hunk 30]
```diff
@@ -295,8 +295,20 @@ def unflatten(s, t, endianness='>'):
...
-FlatData = collections.namedtuple('FlatData', ['bytes', 'tag'])
+class FlatData(collections.namedtuple('FlatDataBase', ['bytes', 'tag', 'endianness'])):
+    """FlatData represents unflattened data.
 
+    Attributes:
+
+    bytes:      The data as intended to be transmitted over the wire
+    tag:        The LabRAD data type in the form of an LRType object.
+    endianness: The endiannes of the connection. '>' for big endian
+                and '<' for little endian.  Little endian is deprecated.
```
Comment (maffoo) [inline labrad/types.py]: nit: docs of individual attributes should be indented under the `Attributes:` section.

File: labrad/decorators.py [hunk 31]
```diff
@@ -21,202 +21,167 @@
...
+
+    def __get__(self, obj, objtype):
+        """Allow direct call on settings from within the Server code
+
+        Uses the python descriptor protocol.
+        """
+        # When called on an instance, we return the setting function
+        # to allow direct calling.  When called on the class we return
+        # the Setting object so that the server code can call
+        # handleRequest
```
Comment (maffoo) [inline labrad/decorators.py]: We should add a test that this works, e.g. a new setting in test_server.py that we call from one of the tests.

**Comment by maffoo** [comment]
A couple of formatting nits and would be good to have a test that server settings are callable as plain methods from within a server. Otherwise LGTM.
